chain/subscriber.go

```diff
@@ -0,0 +1,531 @@
+	return known
+}
+
+func (s *Subscriber) revertChainUpdate(tx ChainUpdateTx, cru chain.RevertUpdate) (err error) {
+	// v1 contracts
+	cru.ForEachFileContractElement(func(fce types.FileContractElement, rev *types.FileContractElement, resolved, valid bool) {
+		if err != nil {
+			return // error occurred
+		}
+
+		var prev, curr *revision
+		if rev != nil {
+			curr = &revision{
+				revisionNumber: rev.FileContract.RevisionNumber,
+				fileSize:       rev.FileContract.Filesize,
+			}
+		}
+		prev = &revision{
+			revisionNumber: fce.FileContract.RevisionNumber,
+			fileSize:       fce.FileContract.Filesize,
+		}
+		err = s.updateContract(tx, cru.State.Index, types.FileContractID(fce.ID), prev, curr, resolved, valid)
+	})
+	if err != nil {
+		return fmt.Errorf("failed to revert v1 contract: %w", err)
+	}
+
+	// v2 contracts
+	cru.ForEachV2FileContractElement(func(fce types.V2FileContractElement, rev *types.V2FileContractElement, res types.V2FileContractResolutionType) {
+		if err != nil {
+			return // error occurred
+		}
+
+		prev := &revision{
+			revisionNumber: fce.V2FileContract.RevisionNumber,
+			fileSize:       fce.V2FileContract.Filesize,
+		}
+		var curr *revision
+		if rev != nil {
+			curr = &revision{
+				revisionNumber: rev.V2FileContract.RevisionNumber,
+				fileSize:       rev.V2FileContract.Filesize,
+			}
+		}
+
+		resolved, valid := checkFileContract(fce, res)
+		err = s.updateContract(tx, cru.State.Index, types.FileContractID(fce.ID), prev, curr, resolved, valid)
+	})
+	if err != nil {
+		return fmt.Errorf("failed to revert v2 contract: %w", err)
+	}
+
+	return nil
+}
+
+func (s *Subscriber) sync(index types.ChainIndex) error {
+	for index != s.cm.Tip() {
+		// check if subscriber was closed
+		select {
+		case <-s.closedChan:
+			return errClosed
+		default:
+		}
+
+		// fetch updates
+		crus, caus, err := s.cm.UpdatesSince(index, updatesBatchSize)
+		if err != nil {
+			return fmt.Errorf("failed to fetch updates: %w", err)
+		}
+
+		// start retry loop
+		for i := 1; i <= len(s.retryTxIntervals)+1; i++ {
+			// check if subscriber was closed
+			select {
+			case <-s.closedChan:
+				return errClosed
+			default:
+			}
+
+			// process updates
+			index, err = s.processUpdates(crus, caus)
+			if err == nil {
+				break
+			}
+
+			// no more retries left
+			if i-1 == len(s.retryTxIntervals) {
+				s.logger.Error(fmt.Sprintf("transaction attempt %d/%d failed, err: %v", i, len(s.retryTxIntervals)+1, err))
+				return fmt.Errorf("failed to process updates after %d attempts: %w", i, err)
+			}
+
+			// sleep
+			interval := s.retryTxIntervals[i-1]
+			s.logger.Warn(fmt.Sprintf("transaction attempt %d/%d failed, retry in %v, err: %v", i, len(s.retryTxIntervals)+1, interval, err))
+			time.Sleep(interval)
+		}
+	}
+	return nil
+}
+
+func (s *Subscriber) processUpdates(crus []chain.RevertUpdate, caus []chain.ApplyUpdate) (index types.ChainIndex, _ error) {
+	// begin a new chain update
+	tx, err := s.cs.BeginChainUpdateTx()
```

Comment:
This would probably be cleaner as a callback so the rollback and commit logic can be put back into the
store.

```
UpdateChainState(fn func(ChainUpdateTx) error) error
```